src/Shared/DiagnosticSourceInstrumentation/PropertyFetcher.cs

```diff
@@ -85,54 +73,109 @@ public bool TryFetch(object obj, out T value, bool skipObjNullCheck = false)
     }
 
     // see https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs
+#if NET6_0_OR_GREATER
+    [RequiresUnreferencedCode(TrimCompatibilityMessage)]
```

Comment:
If you have marked the class `PropertyFetcher` with `RequiresUnreferencedCode` attribute, do you still need to individually add this attribute to other nested types and methods in this class?

Reply:
Or do you even have to add the annotation to the class when you have already targeted the suppression to the specific methods and properties?

Reply:
Yes, we will still need this attribute in this class. We'll need to annotate the call chain.

Reply:
> We'll need to annotate the call chain.

If we just need to annotate the call chain, then we should add annotations ONLY to the specific methods of the type which are not trim safe. Why do we need to annotate the type?

What's the guiding principle here for adding annotations?
- Do we want to add annotations ONLY to the specific methods of a type to clearly demonstrate which code is unsafe?
- Do we want to add the annotation to the type? Does this mean that every method of the type would then be considered unsafe?
- Is there some value to annotating both the type and its methods?

Reply:
> If you have marked the class PropertyFetcher with RequiresUnreferencedCode attribute, do you still need to individually add this attribute to other nested types and methods in this class?

Yes, you need to add the attribute to nested types as well. The `RequiresUnreferencedCode` attribute on an outer class doesn't apply to nested classes. So they both are necessary.

> What's the guiding principle here for adding annotations?

If there are many methods in a class (especially a public class), and only a subset of the methods are incompatible, then just annotating the incompatible methods is preferred.

Annotating the whole class as `RequiresUnreferencedCode` is just a convenience. It means "this whole class can be considered incompatible with trimming. Don't warn inside this class, but just warn all the callers of this class (any constructor calls, and any static method invocations)." You can accomplish the same thing manually by adding the attribute to every method.

In this specific case, `PropertyFetch` only has 2 methods - both of which are incompatible. It wouldn't make sense to add the attribute to both methods when just applying the attribute to the whole class will suffice.

Another situation where using `RequiresUnreferencedCode` at the class level has an advantage is when you have an inheritance hierarchy - like a strategy pattern - and only one of the implementations are incompatible.

```C#

abstract class Abstraction { abstract void Method(); }

class SafeImpl { override void Method() {} }

class UnsafeImpl { override void Method() { // something not compatible with trimming} }
```

In this case, you can't annotate `UnsafeImpl.Method()` as `RequiresUnreferencedCode` because the base method isn't annotated as `RequiresUnreferencedCode`. The whole hierarchy needs to be annotated the same way. And you don't want to make `SafeImpl` raise warnings, because there are no issues with that implementation.

So to solve this, you annotate the whole `UnsafeImpl` class as `RequiresUnreferencedCode`. This warns at all the places that create an `UnsafeImpl` object. And it also resolves the warnings inside `UnsafeImpl.Method()` because the whole class is annotated.